compiler/src/dotty/tools/dotc/ast/TreeInfo.scala

```diff
@@ -578,8 +578,9 @@ trait TypedTreeInfo extends TreeInfo[Type] { self: Trees.Instance[Type] =>
       sym.owner.isPrimitiveValueClass
       || sym.owner == defn.StringClass
       || defn.pureMethods.contains(sym)
+
     tree.tpe.isInstanceOf[ConstantType] && tree.symbol != NoSymbol && isKnownPureOp(tree.symbol) // A constant expression with pure arguments is pure.
-    || fn.symbol.isStableMember && !fn.symbol.is(Lazy)  // constructors of no-inits classes are stable
+    || fn.symbol.isStableMember && fn.symbol.isConstructor // constructors of no-inits classes are stable
```

Comment:
I was a bit confused by what was going on here, but after reading https://github.com/lampepfl/dotty/commit/471fcd9e198204f37995f56c4eb4c5ec10122add it seems the issue is that:
1. Constructors of traits that do not have side-effects will return true to isStableMember
2. Other members might return true just because they're stable (always return the same value) even if they do have side-effects. The previous check excluding lazy was not broad enough to handle this (in fact, arbitrary methods can be marked stable with `scala.annotation.unchecked.uncheckedStable`)

If I understand https://github.com/lampepfl/dotty/commit/471fcd9e198204f37995f56c4eb4c5ec10122add correctly, I think this check is equivalent to:
```suggestion
    || fn.symbol.isConstructor && fn.symbol.owner.is(NoInits)
```
Which is more obviously correct. @sjrd: Can you confirm my understanding here?

Reply:
> I think this check is equivalent to:

I don't think so, because the condition required to set the constructor as stable is not just `fn.symbol.owner.is(NoInits)` but

```scala
if myFlags.is(Trait) then myFlags.is(NoInits) else isNoInitsRealClass
```

where

```scala
    def isNoInitsRealClass(using Context): Boolean =
      isRealClass &&
      (asClass.baseClasses.forall(_.is(NoInits)) || defn.isAssuredNoInits(symbol))
```

So `isStableMember` is the efficient way to verify that.

Perhaps we can `fn.symbol.isStableMember && fn.symbol.isConstructor` into a `isStableConstructor` method, and then throw this detail in its docs.  Or we can do that in situ here.